blockchain/blockindex.go

```diff
@@ -574,14 +869,418 @@ func (bi *blockIndex) SetStatusFlags(node *blockNode, flags blockStatus) {
+	bi.Lock()
+	bi.setStatusFlags(node, statusValidateFailed)
+	bi.removeBestChainCandidate(node)
+	bi.maybeUpdateBestInvalid(node)
+	delete(bi.unlinkedChildrenOf, node)
+
+	// Mark all descendants of the failed block as having a failed ancestor.
+	//
+	// In order to fairly efficiently determine all of the descendants of the
+	// block without having to iterate the entire block index, walk through all
+	// of the known chain tips and check if the block being invalidated is an
+	// ancestor of the tip.  In the case it is, then all blocks between that tip
+	// and the failed block are descendants.  As an additional optimization, a
+	// cache of recent tips (those after a recent height) is maintained and used
+	// when possible to reduce the number of potential affected chain tips that
+	// need to be iterated.
+	//
+	// In order to help visualize the logic, consider the following block tree
+	// with several branches:
+	//
+	// 100 -> 101  -> 102  -> 103  -> 104  -> 105  -> 106  -> 107  -> 108
+	//    \-> 101a -> 102a -> 103a -> 104a -> 105a        \-> 107a
+	//    \-> 101b    ----        |       \-> 105b -> 106b
+	//                 ^^         \-> 104c -> 105c -> 106c
+	//               Failed       \-> 104d -> 105d
+	//
+	// Further, assume block 102a failed validation.  As can be seen, its
+	// descendants are 103a, 104a, 105a, 105b, 106b, 104c, 105c, 106c, 104d, and
+	// 105d, and the chain tips of this hypothetical block tree would be 101b,
+	// 105a, 105d, 106b, 106c, 107a, and 108.
+	//
+	// Since the failed block, 102a, is not an ancestor of tips 101b, 107a, or
+	// 108, those tips are ignored.  Also notice that, of the remaining tips,
+	// 103a is a common ancestor to all of them, and 104a is a common ancestor
+	// to tips 105a and 106b.
+	//
+	// Given all of the above, the blocks would semantically be marked as having
+	// an invalid ancestor as follows:
+	//
+	// Tip 105a: 105a, 104a, 103a       (102a is failed block, next)
+	// Tip 106b: 106b, 105b, 104a, 103a (102a is failed block, next)
+	// Tip 106c: 106c, 105c, 104c, 103a (102a is failed block, next)
+	// Tip 105d: 105d, 104d, 103a       (102a is failed block, next)
+	//
+	// Note that it might be tempting to consider trying to optimize this to
+	// skip directly to the next tip once a node with a common invalid ancestor
+	// is found.  However, that would result in incorrect behavior if a block is
+	// marked invalid deeper in a branch first and then an earlier block is
+	// later marked invalid as it would result in skipping the intermediate
+	// blocks thereby NOT marking them as invalid as they should be.
+	//
+	// For example, consider what happens if 105c is marked invalid prior to the
+	// block data for 102a becoming available and found to be invalid.  Blocks
+	// 106c and 105c would already be marked invalid, and blocks 104c and 103a
+	// need to be marked invalid.
+	markDescendantsFailed := func(node, tip *blockNode) {
```

Comment:
nit: Since this function flags descendant nodes as `statusInvalidAncestor`, would it be worth calling this `markDescendantsInvalid` rather than `markDescendantsFailed `?

Reply:
Yeah, I agree that is more accurate name.  Updated.